src/dataflow/flow-graph.ts

```diff
@@ -0,0 +1,170 @@
+    }
+    return connections;
+  }
+}
+
+/** Creates a new node for every given particle. */
+function createParticleNodes(particles: Particle[]) {
+  const nodes: Map<Particle, ParticleNode> = new Map();
+  particles.forEach(particle => {
+    nodes.set(particle, new ParticleNode(particle));
+  });
+  return nodes;
+}
+
+/** Creates a new node for every given handle. */
+function createHandleNodes(handles: Handle[]) {
+  const nodes: Map<Handle, HandleNode> = new Map();
+  handles.forEach(handle => {
+    nodes.set(handle, new HandleNode(handle));
+  });
+  return nodes;
+}
+
+/** Adds a connection between the given particle and handle nodes. */
+function addHandleConnection(particleNode: ParticleNode, handleNode: HandleNode, connection: HandleConnection) {
+  switch (connection.direction) {
+    case 'in': {
+      const edge = new ParticleInput(particleNode, handleNode, connection.name);
+      particleNode.inEdges.push(edge);
+      handleNode.outEdges.push(edge);
+      break;
+    }
+    case 'out': {
+      const edge = new ParticleOutput(particleNode, handleNode, connection.name);
+      particleNode.outEdges.push(edge);
+      handleNode.inEdges.push(edge);
+      break;
+    }
+    case 'inout': // TODO: Handle inout directions.
+    case 'host':
+    default:
+      throw new Error(`Unsupported connection type: ${connection.direction}`);
+  }
+}
+
+abstract class Node {
+  abstract readonly inEdges: Edge[];
+  abstract readonly outEdges: Edge[];
+
+  get inNodes(): Node[] {
+    return this.inEdges.map(e => e.start);
+  }
+
+  get outNodes(): Node[] {
+    return this.outEdges.map(e => e.end);
+  }
+}
+
+interface Edge {
+  readonly start: Node;
+  readonly end: Node;
+  readonly label: string;
+}
+
+class ParticleNode extends Node {
+  readonly inEdges: ParticleInput[] = [];
+  readonly outEdges: ParticleOutput[] = [];
+  readonly name: string;
+
+  constructor(particle: Particle) {
+    super();
+    this.name = particle.name;
+  }
+}
+
+class ParticleInput implements Edge {
+  readonly start: Node;
+  readonly end: ParticleNode;
+  readonly label: string;
+
+  constructor(particleNode: ParticleNode, otherEnd: Node, inputName: string) {
+    this.start = otherEnd;
+    this.end = particleNode;
+    this.label = `${particleNode.name}.${inputName}`;
+  }
+}
+
+class ParticleOutput implements Edge {
+  readonly start: ParticleNode;
+  readonly end: Node;
+  readonly label: string;
+
+  constructor(particleNode: ParticleNode, otherEnd: Node, outputName: string) {
+    this.start = particleNode;
+    this.end = otherEnd;
+    this.label = `${particleNode.name}.${outputName}`;
+  }
+}
+
+class HandleNode extends Node {
+  readonly inEdges: ParticleOutput[] = [];
+  readonly outEdges: ParticleInput[] = [];
+
+  constructor(handle: Handle) {
+    super();
+  }
+
+  /** Returns a list of all pairs of particles that are connected through this handle, in string form. */
+  get connections(): string[] {
```

Comment:
is this for debugging or for reporting? connectionsAsString or similar?

Reply:
done